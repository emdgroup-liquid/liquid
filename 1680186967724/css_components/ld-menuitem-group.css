:host{display:block}.ld-menuitem-group__list{list-style:none;margin:0;padding:0}.ld-menuitem-group:not(:first-of-type),:host(:not(:first-of-type)){border-top:solid var(--ld-sp-2) var(--ld-col-neutral-050);margin-top:var(--ld-sp-12);padding-top:var(--ld-sp-12)}