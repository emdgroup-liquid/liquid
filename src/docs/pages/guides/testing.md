---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

