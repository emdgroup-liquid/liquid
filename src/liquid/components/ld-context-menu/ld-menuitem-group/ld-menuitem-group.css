:host {
  display: block;
}

.ld-menuitem-group__list {
  list-style: none;
  margin: 0;
  padding: 0;
}

:host(:not(:first-of-type)),
.ld-menuitem-group:not(:first-of-type) {
  border-top: solid var(--ld-sp-2) var(--ld-col-neutral-050);
  margin-top: var(--ld-sp-12);
  padding-top: var(--ld-sp-12);
}
