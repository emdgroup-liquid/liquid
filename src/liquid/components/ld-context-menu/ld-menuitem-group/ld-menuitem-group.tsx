import { Component, Host, h, Prop } from '@stencil/core'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part list - `ul` element wrapping the default slot
 * @part listitem - `li` element wrapping the `ul` element
 */
@Component({
  tag: 'ld-menuitem-group',
  styleUrl: 'ld-menuitem-group.css',
  shadow: true,
})
export class LdMenuitemGroup {
  /** Label for the menu item group. */
  @Prop() ariaLabel: HTMLUListElement['ariaLabel']

  render() {
    return (
      <Host>
        <li class="ld-menuitem-group" part="listitem" role="none">
          <ul
            class="ld-menuitem-group__list"
            aria-label={this.ariaLabel}
            role="group"
          >
            <slot></slot>
          </ul>
        </li>
      </Host>
    )
  }
}
