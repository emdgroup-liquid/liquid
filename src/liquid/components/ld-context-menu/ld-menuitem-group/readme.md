---
eleventyNavigation:
  key: Menuitem Group
  parent: Context Menu
layout: layout.njk
title: Menuitem Group
permalink: components/ld-context-menu/ld-menuitem-group/
---

# ld-menuitem-group

The `ld-menuitem-group` component is a subcomponent for `ld-context-menu`.

Please refer to the [`ld-context-menu` documentation](components/ld-context-menu) for usage examples.

---

<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
