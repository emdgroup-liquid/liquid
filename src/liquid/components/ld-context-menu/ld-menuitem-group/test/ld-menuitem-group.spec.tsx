import { h } from "@stencil/core";
import { newSpecPage } from "@stencil/core/testing";
import { LdMenuitemGroup } from "../ld-menuitem-group";

describe("ld-menuitem-group", () => {
  it("renders", async () => {
    const page = await newSpecPage({
      components: [LdMenuitemGroup],
      template: () => (
        <ld-menuitem-group aria-label="Test label">
          <ld-menuitem>Menu item 1</ld-menuitem>
          <ld-menuitem>Menu item 2</ld-menuitem>
        </ld-menuitem-group>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });
});
