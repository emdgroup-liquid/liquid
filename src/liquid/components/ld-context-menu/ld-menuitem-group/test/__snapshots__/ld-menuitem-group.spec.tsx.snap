// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-menuitem-group renders 1`] = `
<ld-menuitem-group aria-label="Test label">
  <mock:shadow-root>
    <li class="ld-menuitem-group" role="none">
      <ul aria-label="Test label" class="ld-menuitem-group__list" role="group">
        <slot></slot>
      </ul>
    </li>
  </mock:shadow-root>
  <ld-menuitem>
    Menu item 1
  </ld-menuitem>
  <ld-menuitem>
    Menu item 2
  </ld-menuitem>
</ld-menuitem-group>
`;
